Fix TypeUse

kariya-mitsuru opened the pull request:
Inline parameter and result declaration without typeidx should insert
type definition at the last of module.

Inline parameter and result declaration with typeidx should check
consistency.

Passed new 10 tests.

**before**
```
End ".../wain/spec-test/wasm-testsuite/func.wast":
  total: 130, passed: 124, failed: 6, skipped: 0

End ".../wain/spec-test/wasm-testsuite/call_indirect.wast":
  total: 153, passed: 149, failed: 4, skipped: 0

Results of 76 files:
  total: 19757, passed: 19416, failed: 212, skipped: 129
```

**after**
```
End ".../wain/spec-test/wasm-testsuite/func.wast":
  total: 130, passed: 130, failed: 0, skipped: 0

End ".../wain/spec-test/wasm-testsuite/call_indirect.wast":
  total: 153, passed: 153, failed: 0, skipped: 0

Results of 76 files:
  total: 19757, passed: 19426, failed: 202, skipped: 129
```

kariya-mitsuru committed:
Fix TypeUse

Inline parameter and result declaration without typeidx should insert
type definition at the last of module.

Inline parameter and result declaration with typeidx should check
consistency.

rhysd commented on wain-syntax-text/src/wat2wasm.rs:
```diff
@@ -295,6 +334,13 @@ impl<'s, T: Transform<'s>> Transform<'s> for Option<T> {
+                .iter()
+                .map(|wat::TypeDef { ty, .. }| ty.clone())
+                .collect(),
+        );
```
```rust
ctx.types.reserve(self.types.len());
for type_def in self.types.iter() {
    ctx.types.push(type_def.ty.clone());
}
```

にしてもらえると，`.collect()` によるアロケーションが不要になりそうです

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -492,6 +494,81 @@ impl<'s> Parser<'s> {
+            results: results.to_vec(),
+        });
+        idx as u32
+    }
```
メソッド定義の前後は1行空白をあけておいていただけますでしょうか

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -1189,85 +1269,15 @@ impl<'s> Parse<'s> for TypeUse<'s> {
-            idx: Index::Num(idx),
+            idx: match idx {
+                Some(idx) => RefOrInline::Reference(idx),
+                None => RefOrInline::Inline(parser.create_inline_typeuse(start, &params, &results)),
```
これって inline というワードの使い方適切でしょうか？暗黙に自動で生成される ID は仕様では fresh id と呼ばれていたので，fresh のほうが良さそうに感じました

kariya-mitsuru replied:
inline は [spec](https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#text-format%E2%91%A0) で使用されている言葉です。
> A `typeuse` may also be replaced entirely by **inline** *parameter* and *result* declarations.

spec 上この構文要素（`typeuse`）では fresh id と言う用語は出てきません。

rhysd replied:
> A typeuse may also be replaced entirely by inline parameter and result declarations.

https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#abbreviations%E2%91%A6

この inline は parameter and result declarations にかかっているので，例えば `(type 0) (param)` の代わりに `param` が直接書かれているのを指して inline と書いてるのではないでしょうか？`typeuse` の index を省略した時を指して inline と書いているようには読めませんでした．

> spec 上この構文要素（typeuse） fresh id と言う用語は出てきません。

すみません，これは僕の勘違いでした．fresh identifier は function, memory, table, global だけですね．

kariya-mitsuru replied:
> この inline は parameter and result declarations にかかっているので，例えば `(type 0) (param)` の代わりに `param` が直接書かれているのを指して inline と書いてるのではないでしょうか？`typeuse` の index を省略した時を指して inline と書いているようには読めませんでした．

それはその通りです。
が、"`(type n)`" と "`(type n)` (`param`)\* (`result`)\*" のパターンと "(`param`)\* (`result`)\*" のパターンを区別する短くて良い単語が他に思い浮かばなかったので、型定義への**参照を含む**パターンを `Reference`（`param` と `result` は外側の `TypeUse` 構造体に含まれているのでそれ自身で有無を区別する必要が無い）、型定義への参照を含まず**インライン宣言のみ**のパターンを `Inline` 、と言う名称にしました。

とは言え、列挙型名の `RefOrInline` も含めてあまり良い名前とは思っていないので（あまり？）、何か良い命名があればさっくり直します。

rhysd replied:
下記に合わせて Implicit というのはどうでしょう？また enum の名前は僕も良いものが思い付かないのですが，明示するものと暗黙に挿入されるものの2種類があるという意味で TypeUseKind とかどうでしょう？

```rust
enum TypeUseKind<'s> {
    Explicit(Index<'s>),
    Implicit(u32),
}
```

rhysd commented on wain-syntax-text/src/wat2wasm.rs:
```diff
@@ -326,20 +372,22 @@ impl<'s> Transform<'s> for wat::ValType {
-                .results
-                .iter()
-                .map(|p| p.ty.transform(ctx))
-                .collect::<Result<'_, _>>()?,
```
問題ないのですが，ここってどうして `params` と `results` を外に移したのでしょう？borrow checker がらみでしょうか？

kariya-mitsuru replied:
あ、すみません、ボロ―チェッカやデータの値確認とかで悪戦苦闘してた際に切り出したりしてたんですが、最終的に元の状態で上手くいくので戻します。

rhysd commented on wain-syntax-text/src/wat2wasm.rs:
```diff
@@ -470,7 +518,7 @@ impl<'s> Transform<'s> for wat::Instruction<'s> {
             wat::InsnKind::Call(idx) => wasm::InsnKind::Call(ctx.resolve_func_idx(idx, start)?),
             wat::InsnKind::CallIndirect(ty) => {
-                wasm::InsnKind::CallIndirect(ctx.resolve_type_idx(ty.idx, start)?)
+                wasm::InsnKind::CallIndirect(ctx.resolve_type_idx(&ty, start)?)
```
ここムーブして渡していただけますでしょうか？そうすれば `resolve_type_idx` は参照で受け取る必要がなく，したがって `resolve_index` も `Index` を参照で受け取る必要がなくなりそうです

rhysd commented on wain-syntax-text/src/wat2wasm.rs:
```diff
@@ -265,6 +302,8 @@ pub fn wat2wasm<'s>(
         local_indices: Indices::new(),
         next_local_idx: 0,
         label_stack: LabelStack::new(source),
+        tentatives: std::mem::replace(&mut parsed.module.tentatives, Vec::new()),
```
```suggestion
        tentatives: std::mem::take(&mut parsed.module.tentatives),
```

意味は同じですが，`take` でよりシンプルになりそうです

rhysd commented on wain-syntax-text/src/parser.rs:
```diff
@@ -248,6 +248,7 @@ struct ParseContext<'s> {
     table_indices: Indices<'s>,
     mem_indices: Indices<'s>,
     global_indices: Indices<'s>,
+    tentatives: Vec<TentativeTypeUse<'s>>,
```
tentative は temp と同じぐらい情報量が無いので，fresh_types または implicit_types とかでお願いできませんでしょうか？複数のソース（parser.rs と wat2wasm.rs）でまたがって使う値なので，なるべく説明的な変数名を付けておきたいです

kariya-mitsuru replied:
用語 fresh は fresh id ではないのでミスリード、用語 implicit も、この時点では暗黙定義するとは限らない（単なる参照の場合がある）ので不適切と考えました。
ちなみに、tentative は C 言語では仮定義の意味で tentative definition と言うように使用します。今回のケースでは、C 言語では仮定義と同様、定義になる場合と単なる参照になる場合の両方があるためニュアンス的に合ってるなと思い、この用語を使用しました。

「何の？」に当たる部分が無いのはご指摘通りなので、tentative_types ではどうでしょうか？（「仮の」のニュアンスが欲しいな、と思いまして。ホントは tentative_type_definitions とかのほうがいいのかもしれませんが、流石に長すぎる気が…）
他に良い用語があればそれでも構いませんし、もちろんやはり fresh か implicit が良ければそちらにします。

rhysd replied:
~完全に好みの問題で申し訳ないですが，unresolved_type_uses が良いです！！~

すみません，誤読してましたが，`tentatives` は自動挿入される type use の添字 → type use の index のマッピングを表すものなんですね．wat のパースが終わった時点で resolve 自体は完了しているという．うーむ

rhysd replied:
`inserted_type_to_index` とかになるのかな… 一周回って kariya さん案の `tentative_types` のほうが簡潔で良い気もしてきました…（将来的にもっと良い名前を思い付いたらリファクタする感じで…）

rhysd replied:
暗黙に指定されているという意味で implicit_types または implicit_type_uses が良いように思います

kariya-mitsuru replied:
メンバ名は `implicit_type_uses`、構造体名も `ImplicitTypeUse` で修正しました。

rhysd commented:
ありがとうございます．

ちなみにこの fresh ID を後で解決しないといけないのって，type use に限らず table とか memory もそうですよね…

rhysd commented on wain-syntax-text/src/wat2wasm.rs:
```diff
@@ -202,24 +206,57 @@ impl<'s> Context<'s> {
+                let wat::FuncType {
+                    params, results, ..
+                } = &self.types[idx as usize];
+                if ty.params.is_empty() && ty.results.is_empty()
+                    || ty.params.len() == params.len()
+                        && ty.results.len() == results.len()
+                        && ty
+                            .params
+                            .iter()
+                            .zip(params.iter())
+                            .all(|(l, r)| l.ty == r.ty)
+                        && ty
+                            .results
+                            .iter()
+                            .zip(results.iter())
+                            .all(|(l, r)| l.ty == r.ty)
```
```suggestion
if ty.params
        .iter()
        .map(|p| p.ty)
        .eq(params.iter().map(|p| p.ty))
        && ty
            .results
            .iter()
            .map(|r| r.ty)
            .eq(results.iter().map(|r| r.ty))
```

`.eq()` を使うとシーケンスの長さをチェックする必要がなくなるので，よりシンプルになりそうです．

kariya-mitsuru replied:
ちなみに、`is_empty()` は省略できないので残しました。

rhysd replied:
あっ，すみません，そうですね．僕の suggestion のほうが間違いです

kariya-mitsuru committed:
Follow Github comments

Use reserve & push instead of map & collect to avoid unnecessary
allocation.
Add blank line arround the create_inline_typeuse method definition.
Revert unnecessary local variable introduction.
Change the parameter type of the resplve_type_idx from reference to
non reference.
Change std::mem::replace to std::mem::take for simplicity.

kariya-mitsuru commented:
> ちなみにこの fresh ID を後で解決しないといけないのって，type use に限らず table とか memory もそうですよね…

上のコメントにもリプライしましたが、`typeuse` では fresh ID は出てきません。
逆に、fresh ID の解決は（まだちゃんと確認してませんが）後である必要が無いと思います。

`typeuse` は以下の点で特殊だと思います。
- 型定義を自動挿入すべきかどうかはモジュールを最後まで見ないとわからない。
- 型定義を自動挿入する必要がある場合は、モジュールの最後に挿入する必要がある。

まぁ2点目は1点目の裏返しですが…

kariya-mitsuru committed:
Follow Github comments

Use more better names for struct & enum.

rhysd reviewed (approved):
:100:

rhysd commented:
いつもありがとうございます！
